sirepo/pkcli/runner.py

```diff
@@ -0,0 +1,222 @@
+
+
+@contextlib.contextmanager
+def _catch_and_log_errors(exc_type, msg, *args, **kwargs):
+    try:
+        yield
+    except exc_type:
+        pkdlog(msg, *args, **kwargs)
+        pkdlog(pkdexc())
+
+
+# Helper to call container.wait in a async/cancellation-friendly way
+async def _container_wait(container):
+    while True:
+        try:
+            return await trio.run_sync_in_worker_thread(
+                functools.partial(container.wait, timeout=_CANCEL_POLL_INTERVAL)
+            )
+        # ReadTimeout is what the documentation says this raises.
+        # ConnectionError is what it actually raises.
+        # We'll catch both just to be safe.
+        except (requests.exceptions.ReadTimeout, requests.exceptions.ConnectionError):
+            pass
+
+
+class _JobStatus(aenum.Enum):
+    NOT_STARTED = 'NOT_STARTED'
+    RUNNING = 'RUNNING'
+    FINISHED = 'FINISHED'
+
+
+class _JobInfo:
+    def __init__(self, jid, run_dir, status):
+        self.jid = jid
+        self.status = status
+        self.run_dir = run_dir
+
+
+class _JobTracker:
+    def __init__(self, nursery):
+        # XX TODO: eventually we'll need a way to stop this growing without
+        # bound, perhaps by clarifying the split in responsibilities between
+        # the on-disk simulation_db versus the in-memory status.
+        self.jobs = {}
+        self._nursery = nursery
+
+    async def start_job(self, jid, config):
+        await self._nursery.start(self._run_job, jid, config)
+
+    async def _run_job(
+            self, jid, run_dir, config, *, task_status=trio.TASK_STATUS_IGNORED
+    ):
+        # XX TODO: there are all kinds of race conditions here, e.g. if the
+        # same jid gets started multiple times in parallel... we need to
+        # revisit this once jids are globally unique, and possibly add
+        # features like "if there is another job running with the same
+        # user/sim/job but a different hash, then auto-cancel that other job"
+        self.jobs[jid] = _JobInfo(jid, run_dir, _JobStatus.RUNNING)
+        try:
+            container = await trio.run_sync_in_worker_thread(
+                functools.partial(
+                    _DOCKER.containers.run,
+                    config['image'],
+                    config['command'],
+                    working_dir=config['working_dir'],
+                    # {host path: {'bind': container path, 'mode': 'rw'}}
+                    volumes=config['volumes'],
+                    name=jid,
+                    auto_remove=True,
+                    detach=True,
+                    init=True,
+                )
+            )
+            task_status.started()
+            # Returns a dict like: {'Error': None, 'StatusCode': 0}
+            # XX TODO but there may be a race condition where if the container
+            # finishes before our call to wait() starts, then it just errors
+            # out with docker.errors.NotFound
+            with _catch_and_log_errors(
```

Comment:
Unfortunately it's not – catch-all handlers in Trio programs need special error handling syntax, because when you have concurrency it's possible for multiple things to go wrong at the same time, and working with "multi-exceptions" is more complicated. This bit of Trio is actually [in flux](https://github.com/python-trio/trio/issues/611) right now and the new system will make it much easier to write this kind of decorator properly, plus currently we don't have any nurseries inside `_catch_and_log_errors`, so I figured we could fix it up later. But we will want to fix it up eventually, and then it'll be Trio-specific.